### ISSUE TITLE
False positive when overriding an async method

### ISSUE THREAD

**Issue by JelleZijlstra** [python/mypy#5125]
```python
from typing import Any, Callable, Awaitable

AnyFunction = Callable[..., Awaitable[Any]]

def retry() -> Callable[[AnyFunction], AnyFunction]:
    def decorator(fn: AnyFunction) -> AnyFunction:
        return fn
    return decorator

class B:
    async def run(self) -> bool:
        return True

class C(B):
    @retry()
    async def run(self) -> bool:
        return True
```
with master produces
```
$ /home/jelle/mypy/venv/bin/python -m mypy untitled.py 
untitled.py:16: error: Signature of "run" incompatible with supertype "B"
```
This is a regression since 0.600.

**Comment by ilevkivskyi** [comment]
OK, this is an actual type error, you should either update the `AnyFunction` to `Callable[..., Coroutine[None, None, Any]]` or use a type variable with current type as a bound. I would recommend the second one, I always use it myself and it gives tighter types, with `AnyFunction = TypeVar('AnyFunction', bound=Callable[..., Awaitable[Any]])` the code passes.

This is a bit verbose, but I think @gvanrossum proposed that we should actually infer the most precise types for async defs. Possible solution is to define an alias like `NativeCoro = Coroutine[None, None, T]`.

**Comment by JelleZijlstra** [comment]
Thanks for clarifying! So the issue is that in the base class `run` is now inferred as (roughly) `Callable[..., Coroutine]`, and in the child class it is the more general type `Callable[..., Awaitable]`.

I feel like this should perhaps be allowed (just like we allow overriding a function with a different Callable), but it doesn't seem like a big deal.

**Comment by hauntsaninja** [comment]
I think mypy is behaving as expected
